File: AtCoder/AtCoder_Improvement_Guidelines/2-3/D - 暑い日々 (Hot days).py
[d,n],*o=[[*map(int,r.split())] for r in open(0)]
xdp={}
for t, in o[:d]:
    dp={}
    for a,b,c in o[d:]:
        if a<=t<=b:
            v=0
            for xc,xv in xdp.items(): v=max(v,xv+abs(xc-c))
            dp[c]=v
    xdp=dp
print(max(dp.values()))

File: AtCoder/AtCoder_Improvement_Guidelines/2-3/D - 暑い日々 (Hot days).cpp
#include<bits/stdc++.h>
using namespace std;
int main(){
  int D,N; cin>>D>>N;
  int T[D]; for(int &Ti:T)cin>>Ti;
  int A[N],B[N],C[N]; for(int i=0;i<N;i++)cin>>A[i]>>B[i]>>C[i];
  map<int,int>d; int v;
  for(int i=0;i<D;i++){
    map<int,int>td;
    for(int j=0;j<N;j++)if(A[j]<=T[i]&&T[i]<=B[j]){
      v=0; for(auto di:d)v=max(v,di.second+abs(di.first-C[j]));
      td[C[j]]=v;
    }
    d=td;
  }
  v=0; for(auto di:d)v=max(v,di.second);
  cout<<v<<endl;
}